crypto/src/key/rschnorr/mod.rs

```diff
@@ -37,170 +35,177 @@ pub enum RistrettoSignatureError {
+        self.pubkey_data.as_ref().encode_to(dest)
     }
 
     fn size_hint(&self) -> usize {
-        self.pubkey_data.as_bytes().size_hint()
+        self.pubkey_data.as_ref().size_hint()
     }
 
     fn using_encoded<R, F: FnOnce(&[u8]) -> R>(&self, f: F) -> R {
-        self.pubkey_data.as_bytes().using_encoded(f)
+        self.pubkey_data.as_ref().using_encoded(f)
     }
 }
 
 impl Decode for MLRistrettoPublicKey {
     fn decode<I: serialization::Input>(input: &mut I) -> Result<Self, serialization::Error> {
         let v = Vec::decode(input)?;
-        RistrettoPublicKey::from_bytes(&v)
+        schnorrkel::PublicKey::from_bytes(&v)
             .map(|r| MLRistrettoPublicKey { pubkey_data: r })
             .map_err(|_| serialization::Error::from("Public Key deserialization failed"))
     }
 }
 
 impl MLRistrettoPublicKey {
     pub fn as_bytes(&self) -> &[u8] {
-        self.pubkey_data.as_bytes()
+        self.pubkey_data.as_ref()
     }
 
     pub fn from_bytes(bytes: &[u8]) -> Result<Self, RistrettoKeyError> {
         let pk =
-            RistrettoPublicKey::from_bytes(bytes).map_err(|_| RistrettoKeyError::InvalidData)?;
+            schnorrkel::PublicKey::from_bytes(bytes).map_err(|_| RistrettoKeyError::InvalidData)?;
         let result = Self::from_native(pk);
         Ok(result)
     }
 
-    pub fn as_native(&self) -> &RistrettoPublicKey {
+    pub fn as_native(&self) -> &schnorrkel::PublicKey {
         &self.pubkey_data
     }
 
-    pub fn from_native(native: RistrettoPublicKey) -> Self {
+    pub fn from_native(native: schnorrkel::PublicKey) -> Self {
         Self {
             pubkey_data: native,
         }
     }
 
     pub fn from_private_key(private_key: &MLRistrettoPrivateKey) -> Self {
-        Self::from_native(RistrettoPublicKey::from_secret_key(&private_key.key_data))
-    }
-
-    #[allow(unused)]
-    pub(crate) fn verify_challenge(
-        &self,
-        signature: &RistrettoSchnorrSignature,
-        challenge: &[u8],
-    ) -> bool {
-        signature.verify_challenge(self.as_native(), challenge)
+        Self::from_native(private_key.key_data.to_public())
     }
 
-    pub(crate) fn verify_message(&self, signature: &RistrettoSchnorrSignature, msg: &[u8]) -> bool {
-        let e = Blake2b32Stream::new().write(msg).finalize();
-        signature.verify_challenge(self.as_native(), &e)
+    pub(crate) fn verify_message(&self, signature: &schnorrkel::Signature, msg: &[u8]) -> bool {
+        // TODO CRYPTOGRAPHY REVIEW
+        let ctx = schnorrkel::signing_context(SIGNATURE_CONTEXT);
+        self.pubkey_data.verify(ctx.bytes(msg), signature).is_ok()
     }
 }
 
 impl std::ops::Add for &MLRistrettoPublicKey {
     type Output = MLRistrettoPublicKey;
     fn add(self, rhs: Self) -> MLRistrettoPublicKey {
-        let result = self.as_native() + rhs.as_native();
-        MLRistrettoPublicKey::from_native(result)
+        let result = self.pubkey_data.as_point() + rhs.pubkey_data.as_point();
```

Comment:
Is this doing proper elliptic curve addition? From my brief look, seems like it. Now I'm wondering whether verifying the sum of sigs will yield the same result as individual verifications. Can you try this? Create sig1 of msg1 with priv_key1, then sig2 with priv_key2, then verify each, then verify msg1 + msg2 with pub_key1 + pub_key2. I wonder whether that'll work. I wonder whether making hashes internally has ruined this.

Reply:
The addition seems fine